src/sinks/util/buffer/loki.rs

```diff
@@ -80,6 +124,41 @@ impl Batch for LokiBuffer {
     }
 
     fn push(&mut self, mut item: Self::Input) -> PushResult<Self::Input> {
+        // We must sort the stream labels to ensure they hash to
+        // the same stream if the label set matches.
+        item.labels.sort_unstable();
+
+        let partition = &item.partition;
+        if self.latest_timestamps.is_none() {
+            self.partition = Some(item.partition.clone());
+            self.latest_timestamps = Some(self.global_timestamps.take(&partition));
+        }
+        let latest_timestamp = self
+            .latest_timestamps
+            .as_ref()
+            .unwrap()
+            .get(&item.labels)
+            .cloned()
+            .unwrap_or(item.event.timestamp);
+        if item.event.timestamp < latest_timestamp {
+            match self.out_of_order_action {
+                OutOfOrderAction::Drop => {
+                    warn!(
+                        msg = "Received out-of-order event; dropping event.",
+                        internal_log_rate_secs = 30
+                    );
+                    return PushResult::Ok(self.is_full());
+                }
+                OutOfOrderAction::RewriteTimestamp => {
+                    warn!(
+                        msg = "Received out-of-order event, rewriting timestamp.",
+                        internal_log_rate_secs = 30
+                    );
+                    item.event.timestamp = latest_timestamp;
+                }
```

Comment:
Can we (optionally?) persist the original timestamp in some secondary field in this case?

Reply:
This can only be done if encoding is JSON. However, users could do the same thing by setting `remove_timestamp` to false. , the downside is an additional timestamp field for every event, but essentially it is a working solution. I think we should not implement this flag, unless there is a demand for it.

Reply:
Ah, yeah, `remove_timestamp ` already takes care of this! Nvm then.